### PR TITLE
feat(net): add tcp simulation

### PR DESCRIPTION
**Pull request by KveinAxel**
add tcp simulation with the similar interface to `tokio::net::tcp`

### CODE REVIEW

File: madsim/src/sim/net/tcp/network.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+            })),
+        }
+    }
+
+    pub(crate) fn update_config(&self, f: impl FnOnce(&mut Config)) {
+        let mut inner = self.inner.lock().unwrap();
+        f(&mut inner.config);
+    }
+
+    pub(crate) fn reset_node(&self, node: NodeId) -> Result<(), String> {
```
Comment (wangrunji0408) [inline madsim/src/sim/net/tcp/network.rs:59]: use `std::io::Error`?

File: madsim/src/sim/net/tcp/network.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    pub(crate) fn unclog_link(&self, src: NodeId, dst: NodeId) {
+        trace!("tcp unclog: {src} -> {dst}");
+        self.inner.lock().unwrap().clogged_link.remove(&(src, dst));
+    }
+
+    pub(crate) fn listen(
+        &self,
+        id: NodeId,
+        addr: SocketAddr,
+    ) -> Result<mpsc::Receiver<(ConnId, ConnId)>, String> {
```
Comment (wangrunji0408) [inline madsim/src/sim/net/tcp/network.rs:118]: use `std::io::Error`?

File: madsim/src/sim/net/tcp/network.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        let (tx, rx) = mpsc::channel(0);
+        let is_inuse = self
+            .inner
+            .lock()
+            .unwrap()
+            .accept
+            .insert(addr, (id, tx))
+            .is_some();
```
Comment (wangrunji0408) [inline madsim/src/sim/net/tcp/network.rs:126]: If address is in use, the old connection will be replaced. It seems not right?

File: madsim/src/sim/net/tcp/network.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        } else {
+            Ok(rx)
+        }
+    }
+
+    pub(crate) async fn connect(
+        &self,
+        src: NodeId,
+        dst: &SocketAddr,
+    ) -> Result<(ConnId, ConnId), String> {
```
Comment (wangrunji0408) [inline madsim/src/sim/net/tcp/network.rs:138]: use `std::io::Error`?

File: madsim/src/sim/net/tcp/network.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            let mut inner = self.inner.lock().unwrap();
+
+            let (dst, tx) = if let Some((id, tx)) = inner.accept.get_mut(dst) {
+                (*id, tx.clone())
+            } else {
+                return Err("addr not be listened".to_string());
+            };
+
+            let send_conn = inner.conn_cnt.fetch_add(1, Ordering::SeqCst);
+            let recv_conn = inner.conn_cnt.fetch_add(1, Ordering::SeqCst);
```
Comment (wangrunji0408) [inline madsim/src/sim/net/tcp/network.rs:149]: reduce atomic operations
```suggestion
            let send_conn = inner.conn_cnt.fetch_add(2, Ordering::SeqCst);
            let recv_conn = send_conn + 1;
```

File: madsim/src/sim/net/tcp/network.rs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        // if move this send call site into above block, the mutex will cross await
+        tx.send((recv_conn, send_conn))
+            .await
+            .map_err(|e| e.to_string())?;
+
+        trace!("tcp connect conn({}, {})", send_conn, recv_conn);
+        Ok((send_conn, recv_conn))
+    }
+
+    pub(crate) fn send(&self, id: &ConnId, msg: Payload) -> Result<usize, String> {
```
Comment (wangrunji0408) [inline madsim/src/sim/net/tcp/network.rs:168]: use `std::io::Error`?

**Review by wangrunji0408 — COMMENTED** [review]
Generally LGTM. Good work!

File: madsim/src/sim/net/tcp/addr.rs [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+///
+/// # Calling
+///
+/// Currently, this trait is only used as an argument to Tokio functions that
+/// need to reference a target socket address. To perform a `SocketAddr`
+/// conversion directly, use [`lookup_host()`](super::lookup_host()).
+///
+/// This trait is sealed and is intended to be opaque. The details of the trait
+/// will change. Stabilization is pending enhancements to the Rust language.
+pub trait ToSocketAddrs: sealed::ToSocketAddrsPriv {}
```
[the remark this answers is not in the document]
Comment (KveinAxel) [reply madsim/src/sim/net/tcp/addr.rs:45]: the trait is public, but tokio doesn't make the method `to_socket_addrs` public.

File: madsim/src/sim/config.rs [hunk 8]
```diff
@@ -16,6 +16,10 @@ pub struct Config {
     /// Network configurations.
     #[serde(default)]
     pub net: net::Config,
+
+    /// Tcp Configurations
+    #[serde(default)]
+    pub tcp: tcp::Config,
```
[the remark this answers is not in the document]
Comment (KveinAxel) [reply madsim/src/sim/config.rs:22]: maybe we need a different simlutation config for tcp and udp?

[the remark this answers is not in the document]
Comment (wangrunji0408) [reply madsim/src/sim/config.rs:22]: Got it. My point is the current 2 items are duplicated with those in net config.

File: madsim/src/sim/net/tcp/addr.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply madsim/src/sim/net/tcp/addr.rs:45]: oh yes

File: madsim/src/sim/net/tcp/listener.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+use super::{sim::TcpSim, to_socket_addrs, TcpStream, ToSocketAddrs};
+use crate::plugin;
+
+/// a simulated TCP socket server, listen for connections
+#[cfg_attr(docsrs, doc(cfg(madsim)))]
+pub struct TcpListener {
+    // fixme: Maybe here no need for the mutex. but we need access
+    // the receiver without the mut and the acess will cross .await
+    receiver: Mutex<mpsc::Receiver<(u64, u64)>>,
```
[the remark this answers is not in the document]
Comment (KveinAxel) [reply madsim/src/sim/net/tcp/listener.rs:15]: ok, I'll use it.

File: madsim/src/sim/net/tcp/network.rs (hunk 6 above)
Comment (KveinAxel) [reply madsim/src/sim/net/tcp/network.rs:168]: ok, I'll change them all.

File: madsim/src/sim/net/tcp/network.rs (hunk 3 above)
Comment (KveinAxel) [reply madsim/src/sim/net/tcp/network.rs:126]: Yes, it's indeed a buggy implement, I'll fix it.

**Pull request merged**